File: EmarsysPredictShop/EmarsysPredictShop/CategoriesViewController.m
//
//  CategoriesViewController.m
//  EmarsysPredictShop
//

@import UIKit;
#import <EmarsysPredictSDK/EmarsysPredictSDK.h>

#import "CategoriesViewController.h"
#import "Cart+EmarsysPredictSDKExtensions.h"
#import "DataSource.h"
#import "RecommendedItem.h"
#import "RecommendedItemsCell.h"
#import "CategoryViewController.h"
#import "ItemDetailViewController.h"

@interface CategoriesViewController ()

@end

@implementation CategoriesViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    // Do any additional setup after loading the view.

    self.categoriesTableView.delegate = self;
    self.categoriesTableView.dataSource = self;
    self.recommendedTableView.delegate = self;
    self.recommendedTableView.dataSource = self;
    self.recommendResults = [[NSMutableArray alloc] init];
    self.categories = [[NSMutableArray<NSString *> alloc] init];
    self.itemsWithTopics =
        [[NSMutableDictionary<id, NSMutableArray<Item *> *> alloc] init];
    self.categories = [[NSMutableArray<NSString *> alloc] init];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    // Deselect all rows
    for (NSIndexPath *indexPath in _categoriesTableView
             .indexPathsForSelectedRows) {
        [self.categoriesTableView deselectRowAtIndexPath:indexPath animated:NO];
    }
    for (NSIndexPath *indexPath in _recommendedTableView
             .indexPathsForSelectedRows) {
        [self.recommendedTableView deselectRowAtIndexPath:indexPath
                                                 animated:NO];
    }

    self.shopTopics = [[DataSource sharedDataSource] categories];

    [self sendRecommend];
}

- (NSString *)tableView:(UITableView *)tableView
titleForHeaderInSection:(NSInteger)section {
    if (tableView == self.categoriesTableView) {
        return @"";
    } else {
        return [_categories objectAtIndex:section];
    }
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    if (tableView == self.categoriesTableView) {
        return 1;
    }
    return self.categories.count;
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section {
    if (tableView == self.categoriesTableView) {
        return self.shopTopics.count;
    }
    return 1;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"ShowCategory"]) {
        CategoryViewController *ctr = segue.destinationViewController;
        ctr.category =
            self.shopTopics[_categoriesTableView.indexPathForSelectedRow.row];
    } else if ([segue.identifier isEqualToString:@"ShowSelectedRecItem"]) {
        ItemDetailViewController *ctr = segue.destinationViewController;
        RecommendedItem *recItem = sender;
        ctr.item = recItem.item;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (tableView == self.categoriesTableView) {
        UITableViewCell *cell =
            [tableView dequeueReusableCellWithIdentifier:@"CategoryCell"];
        NSAssert(cell, @"No cell");
        cell.textLabel.text = self.shopTopics[indexPath.row];
        return cell;
    } else {
        NSString *category = [_categories objectAtIndex:indexPath.section];
        NSMutableArray *categoryItems =
            [self.itemsWithTopics objectForKey:category];

        RecommendedItemsCell *cell = [tableView
            dequeueReusableCellWithIdentifier:@"RecommendedItemCell"];
        NSAssert(cell, @"No cell");

        cell.items = categoryItems;
        cell.RecommendedCollectionView.dataSource = cell;
        cell.RecommendedCollectionView.delegate = cell;
        [cell.RecommendedCollectionView reloadData];

        return cell;
    }
}

// TOPICAL
- (void)sendRecommend {
    [self.recommendResults removeAllObjects];
    [self.categories removeAllObjects];
    [self.itemsWithTopics removeAllObjects];
    [self.recommendedTableView reloadData];

    EMSession *emarsysSession = [EMSession sharedSession];
    EMTransaction *transaction = [[EMTransaction alloc] init];
    NSArray<EMCartItem *> *cartItems = [[Cart sharedCart] convertItems];
    [transaction setCart:cartItems];

    for (int i = 0; i < 10; i++) {
        NSString *logic = [NSString stringWithFormat:@"%@%i", @"HOME_", i];
        EMRecommendationRequest *recommend =
            [[EMRecommendationRequest alloc] initWithLogic:logic];
        recommend.limit = 10;
        recommend.completionHandler =
            ^(EMRecommendationResult *_Nonnull result) {

              // Process result
              if (result.topic) {

                  if (![self.categories containsObject:result.topic]) {
                      [self.categories addObject:result.topic];
                  }
                  if (![self.itemsWithTopics objectForKey:result.topic]) {
                      NSMutableArray<Item *> *newCategoryItems =
                          [[NSMutableArray<Item *> alloc] init];
                      [self.itemsWithTopics setObject:newCategoryItems
                                               forKey:result.topic];
                  }

                  [result.products enumerateObjectsUsingBlock:^(
                                       EMRecommendationItem *_Nonnull obj,
                                       NSUInteger idx, BOOL *_Nonnull stop) {

                    Item *item = [[Item alloc] initWithItem:obj];
                    [self.recommendResults addObject:item];
                    NSMutableArray *categoryItems =
                        [self.itemsWithTopics objectForKey:result.topic];
                    [categoryItems addObject:item];
                  }];
              }

              dispatch_async(dispatch_get_main_queue(), ^{
                [self.recommendedTableView reloadData];
              });
            };
        [transaction recommend:recommend];
    }

    // Firing the transaction. Should be the last call on the page, called only
    // once.
    [emarsysSession sendTransaction:transaction
                       errorHandler:^(NSError *_Nonnull error) {
                         // Some error happened
                         NSLog(@"value: %@", error);
                         return;
                       }];
}

@end

File: EmarsysMobileShopSwift/EmarsysMobileShopSwift/CategoriesViewController.swift
//
//  CategoriesViewController.swift
//  EmarsysMobileShopSwift
//
//

import UIKit
import EmarsysMobile

class CategoriesViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    @IBOutlet weak var categoriesTableView: UITableView!
    @IBOutlet weak var recommendedTableView: UITableView!
    
    var category: String = ""
    var items = [Item]()
    var recommendResults = [Item]()
    var categories = [String]()
    var shopTopics = [String]()
    var itemsWithTopics = [String:[Item]]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        categoriesTableView.delegate = self
        categoriesTableView.dataSource = self
        recommendedTableView.delegate = self
        recommendedTableView.dataSource = self
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        shopTopics = DataSource.sharedDataSource.categories
        sendRecommend()
    }
    
    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if tableView == categoriesTableView {
            return ""
        } else if section < categories.count {
            var category = categories[section]
            let needle = "Root Catalog>"
            if let range = category.rangeOfString(needle) {
                let idx = range.endIndex
                category = category.substringFromIndex(idx)
            }
            return category
        } else {
            return ""
        }
    }
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        if tableView == categoriesTableView {
            return 1
        }
        return categories.count
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if tableView == categoriesTableView {
            return shopTopics.count
        }
        return 1
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        if tableView == categoriesTableView {
            let cell = tableView.dequeueReusableCellWithIdentifier("CategoryCell")
            let label = shopTopics[indexPath.row] as String
            cell!.textLabel?.text = label
            return cell!
        } else {
            let cell = tableView.dequeueReusableCellWithIdentifier("RecommendedItemCell") as! RecommendedItemsCell
            
            if indexPath.section <= categories.count - 1 {
                let category = categories[indexPath.section]
                let categoryItems = itemsWithTopics[category]
                cell.items = categoryItems ?? [Item]()
                cell.recommendedCollectionView.dataSource = cell
                cell.recommendedCollectionView.delegate = cell
                cell.recommendedCollectionView.reloadData()
            }
            
            return cell
        }
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "ShowCategory" {
            let ctr = segue.destinationViewController as! CategoryViewController
            ctr.category = shopTopics[categoriesTableView.indexPathForSelectedRow!.row]
        } else if segue.identifier == "ShowSelectedRecItem" {
            let ctr = segue.destinationViewController as! ItemDetailViewController
            let recommendedItem = sender as! RecommendedItem
            ctr.item = recommendedItem.item
        }
    }
    
    //TOPICAL
    func sendRecommend() {
        let emarsysSession = EMSession.sharedSession()
        let transaction = EMTransaction()
        let cartItems = Cart.sharedCart.convertItems()
        transaction.setCart(cartItems)
        
        recommendResults.removeAll()
        itemsWithTopics.removeAll()
        categories.removeAll()
        
        for index in 0...10 {
            let logic = NSString(format:"%@%i", "HOME_", index) as String
            let recommend = EMRecommendationRequest(logic: logic)
            recommend.limit = 10
            recommend.completionHandler = {result in
                defer {
                    dispatch_async(dispatch_get_main_queue()) {
                        self.recommendedTableView.reloadData()
                    }
                }
                guard let topic = result.topic else { return }
                
                // Process result
                if !topic.isEmpty {
                    if !self.categories.contains(topic) {
                        self.categories.append(topic)
                    }
                    
                    if self.itemsWithTopics[topic] == nil {
                        let newItemCategory = [Item]()
                        self.itemsWithTopics[topic] = newItemCategory
                    }
                    
                    for obj in result.products {
                        let item = Item(withItem: obj)
                        self.recommendResults.append(item)
                        self.itemsWithTopics[topic]?.append(item)
                    }
                }
            }
            transaction.recommend(recommend)
        }
        
        // Firing the EmarsysMobile queue. Should be the last call on the page, called only once.
        emarsysSession.sendTransaction(transaction, errorHandler: {error in
            // Some error happened
            NSLog("value: %@", error)
            return
        })
    }
    
}
